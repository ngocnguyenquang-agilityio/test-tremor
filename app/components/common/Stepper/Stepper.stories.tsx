// Libs
import type { Meta, StoryObj } from "@storybook/react";

// Components
import Stepper from "./Stepper";

const meta = {
  title: "Components/common/Stepper",
  tags: ["autodocs"],
  component: Stepper,
} as Meta<typeof Stepper>;

export default meta;

type Story = StoryObj<typeof meta>;

const steps = ["Product Info", "Media", "Social", "Pricing"];

export const StepperDefault: Story = {
  args: {
    steps: steps,
    currentStep: 2,
    total: steps.length,
  },
};
