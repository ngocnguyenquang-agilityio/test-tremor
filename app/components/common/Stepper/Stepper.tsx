// Components
import StepperBodyItem from "@/components/common/Stepper/StepperBodyItem/StepperBodyItem";
import StepperFirstItem from "@/components/common/Stepper/StepperFirstItem/StepperFirstItem";
import StepperLastItem from "@/components/common/Stepper/StepperLastItem/StepperLastItem";

// Types
import { Step } from "@/types";

interface StepperProps {
  currentStep: number;
  steps: Step[];
  total: number;
  extendClass?: string;
}

const Stepper = ({ currentStep, steps, total, extendClass }: StepperProps) => {
  const isFirstStep = currentStep === 1;
  const isLastStep = currentStep === total;
  const firstStep = steps[0];
  const lastStep = steps[total - 1];
  const bodySteps = steps.slice(1, total - 1);

  const bodyItems = bodySteps.map(step => {
    return (
      <StepperBodyItem
        key={step.content}
        currentStep={currentStep}
        step={step}
      />
    );
  });

  return (
    <ul
      className={`grid grid-cols-${total} pt-6 pb-4 bg-zinc-700 dark:stepper-dark-bg shadow-box-icon-primary rounded-lg ${extendClass}`}>
      {/* First Item */}
      <StepperFirstItem isFirstStep={isFirstStep} step={firstStep} />
      {/* Body Items */}
      {bodyItems}
      {/* Last Item */}
      <StepperLastItem isLastStep={isLastStep} step={lastStep} />
    </ul>
  );
};

export default Stepper;
