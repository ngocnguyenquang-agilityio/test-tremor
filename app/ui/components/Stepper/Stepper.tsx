// Components
import { Text } from "@tremor/react";

// Types
import { Step } from "@/types";

interface StepperProps {
  currentStep: number;
  steps: Step[];
  total: number;
  extendClass?: string;
}

const Stepper = ({ currentStep, steps, total, extendClass }: StepperProps) => {
  const isFirstStep = currentStep === 1;
  const isLastStep = currentStep === total;
  const firstStep = steps[0];
  const lastStep = steps[total - 1];
  const bodySteps = steps.slice(1, total - 1);

  // Fist Item class
  const firstItemRightLineClass = `1234 h-0.5 w-full ${
    isFirstStep ? "bg-zinc-400" : "bg-white"
  }`;

  // Last Item class
  const lastItemLeftLineClass = `h-0.5 w-[50%] ${
    isLastStep ? "bg-white" : "bg-zinc-400"
  }`;
  const lastItemDotClass = isLastStep ? "stepper-dot-active" : "stepper-dot";
  const lastItemTextClass = `stepper-text-content mt-2 ${
    isLastStep ? "text-white" : "text-zinc-400 dark:text-zinc-400"
  }`;

  // Body Item
  const bodyItems = bodySteps.map(step => {
    const stepIndex = step.index;
    const leftLineClass = `h-0.5 w-[50%] ${
      currentStep >= stepIndex ? "bg-white" : "bg-zinc-400"
    }`;
    const dotClass =
      currentStep >= stepIndex ? "stepper-dot-active" : "stepper-dot";
    const rightLineClass = `h-0.5 w-[50%] ${
      currentStep > stepIndex ? "bg-white" : "bg-zinc-400"
    }`;
    const textClass = `stepper-text-content mt-2 ${
      currentStep >= stepIndex
        ? "text-white"
        : "text-zinc-400 dark:text-zinc-400"
    }`;

    return (
      <li key={step.content}>
        <div className="flex items-center w-full h-4">
          {/* Left line */}
          <span className={leftLineClass}></span>
          {/* Dot */}
          <span className={dotClass}></span>
          {/* Right line */}
          <span className={rightLineClass}></span>
        </div>
        <Text
          className={`${textClass} text-xs`}>{`${step.index}. ${step.content}`}</Text>
      </li>
    );
  });

  return (
    <ul
      className={`grid grid-cols-${total} pt-6 pb-4 bg-zinc-700 dark:stepper-dark-bg shadow-box-icon-primary rounded-lg ${extendClass}`}>
      {/* First Item */}
      <li>
        <div className="flex items-center w-full h-4 pl-[50%]">
          <span className="stepper-dot-active"></span>
          <span className={firstItemRightLineClass}></span>
        </div>
        <Text className="stepper-text-content text-xs text-white mt-2">{`${firstStep.index}. ${firstStep.content}`}</Text>
      </li>
      {/* Body Items */}
      {bodyItems}
      {/* Last Item */}
      <li>
        <div className="flex items-center w-full h-4">
          <span className={lastItemLeftLineClass}></span>
          <span className={lastItemDotClass}></span>
        </div>
        <p
          className={
            lastItemTextClass
          }>{`${lastStep.index}. ${lastStep.content}`}</p>
      </li>
    </ul>
  );
};

export default Stepper;
