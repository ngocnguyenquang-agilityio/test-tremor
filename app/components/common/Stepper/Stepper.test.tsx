// Libs
import { render } from "@testing-library/react";

// Components
import Stepper from "./Stepper";

describe("Stepper component testing", () => {
  it("should match snapshot", () => {
    const mockSteps = ["Product Info", "Media", "Social", "Pricing"];
    const { container } = render(<Stepper steps={mockSteps} currentStep={2} />);

    expect(container).toMatchSnapshot();
  });
});
