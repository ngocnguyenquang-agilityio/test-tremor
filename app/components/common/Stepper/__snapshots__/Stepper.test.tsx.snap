// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper component testing should match snapshot 1`] = `
<div>
  <ul
    class="grid grid-cols-4 pt-6 pb-4 bg-zinc-700 dark:stepper-dark-bg shadow-box-icon-primary rounded-lg undefined"
  >
    <li>
      <div
        class="flex items-center w-full h-4 pl-[50%]"
      >
        <span
          class="stepper-dot-active"
        />
        <span
          class="h-0.5 w-full bg-white"
        />
      </div>
      <p
        class="stepper-text-content text-white mt-2"
      >
        1. Product Info
      </p>
    </li>
    <li>
      <div
        class="flex items-center w-full h-4"
      >
        <span
          class="h-0.5 w-[50%] bg-white"
        />
        <span
          class="stepper-dot-active"
        />
        <span
          class="h-0.5 w-[50%] bg-zinc-400"
        />
      </div>
      <p
        class="stepper-text-content mt-2 text-white"
      >
        2. Media
      </p>
    </li>
    <li>
      <div
        class="flex items-center w-full h-4"
      >
        <span
          class="h-0.5 w-[50%] bg-zinc-400"
        />
        <span
          class="stepper-dot"
        />
        <span
          class="h-0.5 w-[50%] bg-zinc-400"
        />
      </div>
      <p
        class="stepper-text-content mt-2 text-zinc-400"
      >
        3. Social
      </p>
    </li>
    <li>
      <div
        class="flex items-center w-full h-4"
      >
        <span
          class="h-0.5 w-[50%] bg-zinc-400"
        />
        <span
          class="stepper-dot"
        />
      </div>
      <p
        class="stepper-text-content mt-2 text-zinc-400"
      >
        4. Pricing
      </p>
    </li>
  </ul>
</div>
`;
